OpenAdaptAI/OpenAdapt
# feat(crud): compute and save screenshot diff

**What kind of change does this PR introduce?**

* Add 2 columns in screenshot table to store png_diff_data and png_mask_diff_data.

* CRUD now supports calculation and save screenshots diff data on the flight.



**Summary**


This PR solves https://github.com/OpenAdaptAI/OpenAdapt/issues/348. The main logic behinds the change is to introduce the function to calculate and then eventually store `png_data` difference between 2 linked screenshots. Motivation for the design:

- 2 columns that store `png_diff_data` and `png_mask_diff_data` is BLOB/binary to conform with existing `png_data`.
- However, the `Screenshot.diff` and `Screenshot.diff_mask` are kept as `PIL Image` for convenience of usage.

Side effects:
- If called, the calculation of screenshot diff is resource intensive for first time calculation, as a pairs of linked screenshot are compared, then converted to binary and store in database.
- The size of the storage significantly increases. As of now, `png_data` is the column that consumes the most space.
- It will makes space-storage worst, which is already limited by the fact that the max length of recording is ~1 minute for now.
- However, those concerns are mitigated by the fact that `precompute_diff` flag is opt-in an can be used only for case where optimization is required.

**Checklist**

**How can your code be run and tested?**
To activate screen diff storage, set `SAVE_SCREENSHOT_DIFF` environment variable to `True`

```bash
export SAVE_SCREENSHOT_DIFF=True
python -m openadapt.replay DemoReplayStrategy
```

This feature can be activated by an environment variable Flag. Discussion in the PR. 

- The first run should take long time as the screenshot diffs are calculated and stored.
- The second run should be instant, as the diff are only loaded from existing database.

Diff stored as BLOB in database:

<img width="897" alt="Screenshot 2023-07-09 at 17 42 07" src="https://github.com/OpenAdaptAI/OpenAdapt/assets/6369285/b40a9d0b-6d08-4be6-a64a-ca368d38d591">
